File: packages/trading/octobot_trading/exchanges/adapters/abstract_adapter.py
#  Drakkar-Software OctoBot-Trading
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import decimal

import octobot_trading.errors as errors
import octobot_trading.enums as enums
import octobot_commons.logging as logging
import octobot_commons.constants as commons_constants


def _adapter(func):
    def adapter_wrapper(*args, **kwargs):
        try:
            if args[1] is None:
                # element to adapt is None, no need to go any further
                return None
            adapted = func(*args, **kwargs)
            # add any other common adapter function logic here
            return adapted
        except Exception as err:
            raise errors.UnexpectedAdapterError(
                f"Unexpected error when adapting exchange data: {err} (data: {args[1]})"
            ) from err
    return adapter_wrapper


class AbstractAdapter:
    def __init__(self, connector):
        self.logger = logging.get_logger(self.__class__.__name__)
        self.connector = connector

    @_adapter
    def adapt_orders(self, raw, cancelled_only=False, **kwargs) -> list[dict]:
        return [
            order
            for order in (
                self.adapt_order(element, **kwargs)
                for element in raw
            )
            if not cancelled_only or not order or (
                order[enums.ExchangeConstantsOrderColumns.STATUS.value] == enums.OrderStatus.CANCELED.value
            )
        ]

    @_adapter
    def adapt_order(self, raw, **kwargs) -> dict:
        fixed = self.fix_order(raw, **kwargs)
        return self.parse_order(fixed, **kwargs)

    @_adapter
    def adapt_ohlcv(self, raw, **kwargs) -> list[dict]:
        fixed = self.fix_ohlcv(raw, **kwargs)
        return self.parse_ohlcv(fixed, **kwargs)

    @_adapter
    def adapt_kline(self, raw, **kwargs) -> dict:
        fixed = self.fix_kline(raw, **kwargs)
        return self.parse_kline(fixed, **kwargs)

    @_adapter
    def adapt_ticker(self, raw, **kwargs) -> dict:
        fixed = self.fix_ticker(raw, **kwargs)
        return self.parse_ticker(fixed, **kwargs)

    @_adapter
    def adapt_ticker_from_kline(self, raw, symbol, **kwargs) -> dict:
        fixed = self.create_ticker_from_kline(raw, symbol, **kwargs)
        return self.parse_ticker(fixed, **kwargs)

    @_adapter
    def adapt_balance(self, raw, **kwargs) -> dict:
        fixed = self.fix_balance(raw, **kwargs)
        return self.parse_balance(fixed, **kwargs)

    @_adapter
    def adapt_order_book(self, raw, **kwargs) -> dict:
        fixed = self.fix_order_book(raw, **kwargs)
        return self.parse_order_book(fixed, **kwargs)

    @_adapter
    def adapt_public_recent_trades(self, raw, **kwargs) -> list[dict]:
        fixed = self.fix_public_recent_trades(raw, **kwargs)
        return self.parse_public_recent_trades(fixed, **kwargs)

    @_adapter
    def adapt_trades(self, raw, **kwargs) -> list[dict]:
        fixed = self.fix_trades(raw, **kwargs)
        return self.parse_trades(fixed, **kwargs)

    @_adapter
    def adapt_position(self, raw, **kwargs) -> dict:
        fixed = self.fix_position(raw, **kwargs)
        return self.parse_position(fixed, **kwargs)

    @_adapter
    def adapt_funding_rate(self, raw, **kwargs) -> dict:
        fixed = self.fix_funding_rate(raw, **kwargs)
        return self.parse_funding_rate(fixed, **kwargs)

    @_adapter
    def adapt_leverage_tiers(self, raw, **kwargs) -> dict[str, list[dict]]:
        fixed = self.fix_leverage_tiers(raw, **kwargs)
        return self.parse_leverage_tiers(fixed, **kwargs) 

    @_adapter
    def adapt_leverage(self, raw, **kwargs) -> dict:
        fixed = self.fix_leverage(raw, **kwargs)
        return self.parse_leverage(fixed, **kwargs)

    @_adapter
    def adapt_funding_rate_history(self, raw, **kwargs) -> list[dict]:
        fixed = self.fix_funding_rate_history(raw, **kwargs)
        return self.parse_funding_rate_history(fixed, **kwargs)

    @_adapter
    def adapt_mark_price(self, raw, **kwargs) -> dict:
        fixed = self.fix_mark_price(raw, **kwargs)
        return self.parse_mark_price(fixed, **kwargs)

    @_adapter
    def adapt_market_status(self, raw, remove_price_limits=False, **kwargs) -> dict:
        fixed = self.fix_market_status(raw, remove_price_limits=remove_price_limits, **kwargs)
        return self.parse_market_status(fixed, remove_price_limits=remove_price_limits, **kwargs)

    @_adapter
    def adapt_transaction(self, raw, **kwargs) -> dict:
        fixed = self.fix_transaction(raw, **kwargs)
        return self.parse_transaction(fixed, **kwargs)

    @_adapter
    def adapt_deposit_address(self, raw, **kwargs) -> dict:
        fixed = self.fix_deposit_address(raw, **kwargs)
        return self.parse_deposit_address(fixed, **kwargs)

    def get_uniformized_timestamp(self, timestamp) -> float:
        # override if the exchange time is not a second timestamp or millisecond
        if timestamp is not None and timestamp > 16728292300:  # Friday 5 February 2500 11:51:40
            return timestamp / commons_constants.MSECONDS_TO_SECONDS
        return timestamp

    def safe_decimal(self, container, key, default) -> decimal.Decimal:
        if (val := container.get(key, default)) is not None:
            return decimal.Decimal(str(val))
        return default

    def fix_order(self, raw, **kwargs) -> dict:
        # id is reserved for octobot managed id. store exchange id in EXCHANGE_ID
        if enums.ExchangeConstantsOrderColumns.ID.value in raw:
            raw[enums.ExchangeConstantsOrderColumns.EXCHANGE_ID.value] = \
                raw.pop(enums.ExchangeConstantsOrderColumns.ID.value, None)
        # add generic logic if necessary
        return raw

    def parse_order(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_order is not implemented")

    def fix_ohlcv(self, raw, **kwargs) -> list[dict]:
        # add generic logic if necessary
        return raw

    def parse_ohlcv(self, fixed, **kwargs) -> list[dict]:
        raise NotImplementedError("parse_ohlcv is not implemented")

    def fix_kline(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_kline(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_kline is not implemented")

    def fix_ticker(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_ticker(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_ticker is not implemented")

    def create_ticker_from_kline(self, kline, symbol, **kwargs) -> dict:
        raise NotImplementedError("create_ticker_from_kline is not implemented")

    def fix_balance(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_balance(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_balance is not implemented")

    def fix_order_book(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_order_book(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_order_book is not implemented")

    def fix_public_recent_trades(self, raw, **kwargs) -> list[dict]:
        # add generic logic if necessary
        return raw

    def parse_public_recent_trades(self, fixed, **kwargs) -> list[dict]:
        raise NotImplementedError("parse_public_recent_trades is not implemented")

    def fix_trades(self, raw, **kwargs) -> list[dict]:
        for trade in raw:
            # id is reserved for octobot managed id. store exchange trade id in EXCHANGE_TRADE_ID
            trade[enums.ExchangeConstantsOrderColumns.EXCHANGE_TRADE_ID.value] = \
                trade.pop(enums.ExchangeConstantsOrderColumns.ID.value, None)
            # add generic logic if necessary
        return raw

    def parse_trades(self, fixed, **kwargs) -> list[dict]:
        raise NotImplementedError("parse_trades is not implemented")

    def fix_position(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_position(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_position is not implemented")

    def fix_funding_rate(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_funding_rate(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_funding_rate is not implemented")

    def fix_leverage(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_leverage(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_leverage is not implemented")

    def fix_funding_rate_history(self, raw, **kwargs) -> list[dict]:
        # add generic logic if necessary
        return raw

    def parse_funding_rate_history(self, fixed, **kwargs) -> list[dict]:
        raise NotImplementedError("parse_funding_rate_history is not implemented")

    def fix_leverage_tiers(self, raw, **kwargs) -> dict[str, list[dict]]:
        # add generic logic if necessary
        return raw

    def parse_leverage_tiers(self, fixed, **kwargs) -> dict[str, list[dict]]:
        raise NotImplementedError("parse_leverage_tiers is not implemented")

    def fix_mark_price(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_mark_price(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_mark_price is not implemented")

    def fix_market_status(self, raw, remove_price_limits=False, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_market_status(self, fixed, remove_price_limits=False, **kwargs) -> dict:
        raise NotImplementedError("parse_market_status is not implemented")

    def fix_transaction(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_transaction(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_transaction is not implemented")

    def fix_deposit_address(self, raw, **kwargs) -> dict:
        # add generic logic if necessary
        return raw

    def parse_deposit_address(self, fixed, **kwargs) -> dict:
        raise NotImplementedError("parse_deposit_address is not implemented")

File: packages/trading/octobot-trading-ts/src/exchanges/adapters/abstractAdapter.ts
import { ExchangeConstantsOrderColumns, OrderStatus } from "../enums.js";
import { UnexpectedAdapterError } from "../errors.js";
import { Logger, getLogger } from "../util/logger.js";
import { toDecimal } from "../util/decimalUtil.js";
import Decimal from "decimal.js";

const MSECONDS_TO_SECONDS = 1000;
const FUTURE_TIMESTAMP_THRESHOLD = 16728292300; // matches commons_constants

// Mirrors the @_adapter decorator from abstract_adapter.py: short-circuits when
// the input is null/undefined and wraps any thrown error in UnexpectedAdapterError.
function adapter<TIn, TOut, A extends unknown[]>(
  fn: (raw: TIn, ...args: A) => TOut,
): (raw: TIn | null | undefined, ...args: A) => TOut | null {
  return function wrapped(this: AbstractAdapter, raw, ...args) {
    if (raw === null || raw === undefined) return null;
    try {
      return fn.call(this, raw, ...args);
    } catch (err) {
      throw new UnexpectedAdapterError(
        `Unexpected error when adapting exchange data: ${(err as Error).message} (data: ${JSON.stringify(raw)})`,
      );
    }
  } as (raw: TIn | null | undefined, ...args: A) => TOut | null;
}

// Adapter inputs/outputs are loose dictionaries from CCXT, so we use generic
// records for the public API. Subclasses can narrow further via TypedDict-style
// interfaces if they want.
export type RawDict = Record<string, unknown>;

export interface AdaptOrdersOptions {
  cancelledOnly?: boolean;
  [key: string]: unknown;
}

export class AbstractAdapter {
  protected logger: Logger;
  // typed loosely on purpose — `Connector` is forward-declared and may be a
  // CCXTConnector or a custom one.
  constructor(public connector: unknown) {
    this.logger = getLogger(this.constructor.name);
  }

  // ---- adapt_*: orchestration shells (decorator + fix → parse) ----

  adaptOrders(raw: RawDict[] | null | undefined, opts: AdaptOrdersOptions = {}): RawDict[] {
    if (raw === null || raw === undefined) return [];
    try {
      const result: RawDict[] = [];
      for (const element of raw) {
        const adapted = this.adaptOrder(element, opts);
        if (!opts.cancelledOnly || !adapted ||
            adapted[ExchangeConstantsOrderColumns.STATUS] === OrderStatus.CANCELED) {
          if (adapted) result.push(adapted);
        }
      }
      return result;
    } catch (err) {
      throw new UnexpectedAdapterError(
        `Unexpected error when adapting orders: ${(err as Error).message}`,
      );
    }
  }

  adaptOrder = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixOrder(raw, opts);
    return this.parseOrder(fixed, opts);
  });

  adaptOhlcv = adapter<RawDict, RawDict[], [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixOhlcv(raw, opts);
    return this.parseOhlcv(fixed, opts);
  });

  adaptKline = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixKline(raw, opts);
    return this.parseKline(fixed, opts);
  });

  adaptTicker = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixTicker(raw, opts);
    return this.parseTicker(fixed, opts);
  });

  adaptTickerFromKline = adapter<RawDict, RawDict, [symbol: string, opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    symbol,
    opts = {},
  ) {
    const fixed = this.createTickerFromKline(raw, symbol, opts);
    return this.parseTicker(fixed, opts);
  });

  adaptBalance = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixBalance(raw, opts);
    return this.parseBalance(fixed, opts);
  });

  adaptOrderBook = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixOrderBook(raw, opts);
    return this.parseOrderBook(fixed, opts);
  });

  adaptPublicRecentTrades = adapter<RawDict[], RawDict[], [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixPublicRecentTrades(raw, opts);
    return this.parsePublicRecentTrades(fixed, opts);
  });

  adaptTrades = adapter<RawDict[], RawDict[], [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixTrades(raw, opts);
    return this.parseTrades(fixed, opts);
  });

  adaptPosition = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixPosition(raw, opts);
    return this.parsePosition(fixed, opts);
  });

  adaptFundingRate = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixFundingRate(raw, opts);
    return this.parseFundingRate(fixed, opts);
  });

  adaptLeverageTiers = adapter<RawDict, Record<string, RawDict[]>, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixLeverageTiers(raw, opts);
    return this.parseLeverageTiers(fixed, opts);
  });

  adaptLeverage = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixLeverage(raw, opts);
    return this.parseLeverage(fixed, opts);
  });

  adaptFundingRateHistory = adapter<RawDict[], RawDict[], [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixFundingRateHistory(raw, opts);
    return this.parseFundingRateHistory(fixed, opts);
  });

  adaptMarkPrice = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixMarkPrice(raw, opts);
    return this.parseMarkPrice(fixed, opts);
  });

  adaptMarketStatus = adapter<RawDict, RawDict, [removePriceLimits?: boolean, opts?: RawDict]>(
    function (this: AbstractAdapter, raw, removePriceLimits = false, opts = {}) {
      const fixed = this.fixMarketStatus(raw, removePriceLimits, opts);
      return this.parseMarketStatus(fixed, removePriceLimits, opts);
    },
  );

  adaptTransaction = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixTransaction(raw, opts);
    return this.parseTransaction(fixed, opts);
  });

  adaptDepositAddress = adapter<RawDict, RawDict, [opts?: RawDict]>(function (
    this: AbstractAdapter,
    raw,
    opts = {},
  ) {
    const fixed = this.fixDepositAddress(raw, opts);
    return this.parseDepositAddress(fixed, opts);
  });

  // ---- helpers ----

  getUniformizedTimestamp(timestamp: number | null | undefined): number | null {
    if (timestamp === null || timestamp === undefined) return null;
    if (timestamp > FUTURE_TIMESTAMP_THRESHOLD) {
      return timestamp / MSECONDS_TO_SECONDS;
    }
    return timestamp;
  }

  safeDecimal(container: RawDict, key: string, defaultValue: Decimal): Decimal {
    const val = container[key];
    if (val !== null && val !== undefined) {
      return toDecimal(val as string);
    }
    return defaultValue;
  }

  // ---- fix_* / parse_* hooks: subclasses override parse_*; fix_* default no-op ----

  fixOrder(raw: RawDict, _opts: RawDict = {}): RawDict {
    if (ExchangeConstantsOrderColumns.ID in raw) {
      raw[ExchangeConstantsOrderColumns.EXCHANGE_ID] = raw[ExchangeConstantsOrderColumns.ID];
      delete raw[ExchangeConstantsOrderColumns.ID];
    }
    return raw;
  }
  parseOrder(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseOrder is not implemented");
  }

  fixOhlcv(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseOhlcv(_fixed: RawDict, _opts: RawDict = {}): RawDict[] {
    throw new Error("parseOhlcv is not implemented");
  }

  fixKline(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseKline(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseKline is not implemented");
  }

  fixTicker(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseTicker(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseTicker is not implemented");
  }

  createTickerFromKline(_kline: RawDict, _symbol: string, _opts: RawDict = {}): RawDict {
    throw new Error("createTickerFromKline is not implemented");
  }

  fixBalance(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseBalance(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseBalance is not implemented");
  }

  fixOrderBook(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseOrderBook(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseOrderBook is not implemented");
  }

  fixPublicRecentTrades(raw: RawDict[], _opts: RawDict = {}): RawDict[] {
    return raw;
  }
  parsePublicRecentTrades(_fixed: RawDict[], _opts: RawDict = {}): RawDict[] {
    throw new Error("parsePublicRecentTrades is not implemented");
  }

  fixTrades(raw: RawDict[], _opts: RawDict = {}): RawDict[] {
    for (const trade of raw) {
      trade[ExchangeConstantsOrderColumns.EXCHANGE_TRADE_ID] = trade[ExchangeConstantsOrderColumns.ID];
      delete trade[ExchangeConstantsOrderColumns.ID];
    }
    return raw;
  }
  parseTrades(_fixed: RawDict[], _opts: RawDict = {}): RawDict[] {
    throw new Error("parseTrades is not implemented");
  }

  fixPosition(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parsePosition(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parsePosition is not implemented");
  }

  fixFundingRate(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseFundingRate(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseFundingRate is not implemented");
  }

  fixLeverage(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseLeverage(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseLeverage is not implemented");
  }

  fixFundingRateHistory(raw: RawDict[], _opts: RawDict = {}): RawDict[] {
    return raw;
  }
  parseFundingRateHistory(_fixed: RawDict[], _opts: RawDict = {}): RawDict[] {
    throw new Error("parseFundingRateHistory is not implemented");
  }

  fixLeverageTiers(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseLeverageTiers(_fixed: RawDict, _opts: RawDict = {}): Record<string, RawDict[]> {
    throw new Error("parseLeverageTiers is not implemented");
  }

  fixMarkPrice(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseMarkPrice(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseMarkPrice is not implemented");
  }

  fixMarketStatus(raw: RawDict, _removePriceLimits = false, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseMarketStatus(_fixed: RawDict, _removePriceLimits = false, _opts: RawDict = {}): RawDict {
    throw new Error("parseMarketStatus is not implemented");
  }

  fixTransaction(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseTransaction(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseTransaction is not implemented");
  }

  fixDepositAddress(raw: RawDict, _opts: RawDict = {}): RawDict {
    return raw;
  }
  parseDepositAddress(_fixed: RawDict, _opts: RawDict = {}): RawDict {
    throw new Error("parseDepositAddress is not implemented");
  }
}
